service/interpreter/stateExecutionCounter.go

```diff
@@ -84,15 +97,17 @@ func (e *StateExecutionCounter) MarkStateIdExecutingIfNotYet(stateReqs []StateRe
 			case iwfidl.DISABLED:
 				// do nothing
 			case iwfidl.ENABLED_FOR_ALL:
-				if e.increaseStateIdCurrentlyExecutingCounts(s) {
+				e.increaseStateIdCurrentlyExecutingCounts(s)
+				if !slices.Contains(currentSAs.StringArrayValue, s.StateId) {
```

Comment:
After reflecting into this for a while, I found issues with this comparison. Just checking containing is probably not correct.

Consider this case (assuming using ENABLED_FOR_ALL so that we don't care about the waitUntil):
* S1->S2+S1
* S2->S1
So there are 3 S1 executions. Assuming the first S1-1 completed, then goes to S2-1 and S1-2. But S1-2 didn't complete yet, then S2-1 completed and decided to go to S1-3.
Now S1-3 will call this `MarkStateIdExecutingIfNotYet`:
* Because of the optimization in `MarkStateExecutionCompleted`, `S2` is not deleted and still in the search attribute. So the SA contains both: [S1, S2]. (Assuming you are implementing the optimization as we discussed today -- skip deleting the stateID from SA if there is a next state going to refresh the whole SA)
* Now for S1-3 calling `MarkStateIdExecutingIfNotYet`, it will see the SA contains S1 already, but it also contains S2. Therefore, it should still refresh the SA based on `stateIdCurrentlyExecutingCounts` (Note: it may be a lot better to rename `updateStateIdSearchAttribute` -> `refreshIwfExecutingStateIdSearchAttribute` to be more clear.

I remember you were doing the full comparison when you put it in the `updateStateIdSearchAttribute`, not sure why you changed your mind 🤣 

I think you may still run into NDE if putting it into that method, but instead, maybe you can create another method `checkIfNeedToRefreshIwfExecutingStateIdSearchAttribute` before calling `refreshIwfExecutingStateIdSearchAttribute`.